asmcomp/cmm_helpers.ml

```diff
@@ -1410,24 +1513,40 @@ let make_switch arg cases actions dbg =
     Cmmgen_state.add_constant table (Const_table (Local,
         Array.to_list (Array.map (fun act ->
           const_actions.(act)) cases)));
-    addr_array_ref (Cconst_symbol (table, dbg)) (tag_int arg dbg) dbg
+    (* Constant integers loaded from a table are tagged,
+       so that Cload never produces untagged integers. *)
+    addr_array_ref (Cconst_symbol (table, dbg)) arg dbg
```

Comment:
Honestly, this could be `int_array_ref` instead, and we wouldn't need to care about what `Cload` produces. I don't mind if you don't feel like changing that in this PR though.

Reply:
The content of the constant table is what `extract_uconstant` accepts, so it can be either a tagged integer or a `Csymbol_address s` value -- I assume that this could be, for example, the address of a string literal. Are you sure that we can use `int_array_ref` in this case? I'll trust you on that.

Reply:
The reasoning is unconventional. `Csymbol_address` returns a pointer to an OCaml block, so by default it should be of machtype `typ_val`. But this block is not in the heap, and is guaranteed to be already registered as a root if it may contain values, so there's no need to actually treat it as a root at runtime. Hence, `typ_int` works, and is actually very slightly more performant.

Because the reasoning is complex, and the gains negligible, I don't think it makes sense to actually make this change, but if removing that restriction would have enabled more optimisations I might have pushed it.
I'm resolving the conversation to indicate that I'm not expecting further changes in this PR.